### PR TITLE
Remove storage cache layer (#324)

### PR DESCRIPTION
**Pull request by codingllama**
The following changes have been made:

* Open *sql.DB outside of storage (and reuse the same DB for all storage objects)
* Test refactors (storage/mysql):
  * Moved DB-handling methods to storage_test (which contains TestMain)
  * Added a *sql.DB parameter to methods that access storage
  * Open DB connection only once for all tests
  * Cleaned DB at the beginning of every test

The changes above render the storage caching layer obsolete.

### CODE REVIEW

File: storage/mysql/storage_test.go [hunk 1]
```diff
@@ -166,9 +157,83 @@ func nodesAreEqual(lhs []storage.Node, rhs []storage.Node) error {
...
+func openTestDBOrDie() *sql.DB {
+	db, err := OpenDB("test:zaphod@tcp(127.0.0.1:3306)/test")
+	if err != nil {
+		panic(err)
+	}
+	return db
+}
+
+// cleanTestDB deletes all the entries in the database. Only use this with a test database
+func cleanTestDB(db *sql.DB) {
```
Comment (gdbelvin) [inline storage/mysql/storage_test.go]: Consider using a variant of the test specific new db logic in `testonly/integration` which creates a separate database for each test.

File: storage/mysql/storage_test.go [hunk 2]
```diff
@@ -166,9 +157,83 @@ func nodesAreEqual(lhs []storage.Node, rhs []storage.Node) error {
...
+	for _, table := range allTables {
+		_, err := db.Exec(fmt.Sprintf("DELETE FROM %s", table))
+
+		if err != nil {
+			panic(fmt.Errorf("Failed to delete rows in %s: %s", table, err))
+		}
+	}
+}
+
+func createTestDB(db *sql.DB) {
```
Comment (gdbelvin) [inline storage/mysql/storage_test.go]: Consider using the temporary tree admin api in `storage/mysql/log_admin.go`

File: storage/mysql/storage_test.go [hunk 3]
```diff
@@ -166,9 +157,83 @@ func nodesAreEqual(lhs []storage.Node, rhs []storage.Node) error {
...
+	_, err := db.Exec(`REPLACE INTO Trees(TreeId, KeyId, TreeType, LeafHasherType, TreeHasherType)
+					 VALUES(23, "hi", "LOG", "SHA256", "SHA256")`)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// prepareTestTreeDB removes all database contents for the specified log id so tests run in a predictable environment. For obvious reasons this should only be allowed to run against test databases. This method panics if any of the deletions fails to make sure tests can't inadvertently succeed.
+func prepareTestTreeDB(db *sql.DB, treeID int64, t *testing.T) {
```
Comment (gdbelvin) [inline storage/mysql/storage_test.go]: Consider deleting the existing test database and creating a new one for test freshness.

File: storage/mysql/storage_test.go [hunk 4]
```diff
@@ -166,9 +157,83 @@ func nodesAreEqual(lhs []storage.Node, rhs []storage.Node) error {
...
+		_, err := db.Exec(fmt.Sprintf("DELETE FROM %s WHERE TreeId=?", table), treeID)
+
+		if err != nil {
+			t.Fatalf("Failed to delete rows in %s for %d: %s", table, treeID, err)
+		}
+	}
+}
+
+// prepareTestLogDB removes all database contents for the specified log id so tests run in a predictable environment. For obvious reasons this should only be allowed to run against test databases. This method panics if any of the deletions fails to make sure tests can't inadvertently succeed.
+func prepareTestLogDB(db *sql.DB, logID logIDAndTest, t *testing.T) {
```
Comment (gdbelvin) [inline storage/mysql/storage_test.go]: Consider using the tree admin api

File: storage/mysql/storage_test.go (hunk 1 above)
Comment (codingllama) [inline storage/mysql/storage_test.go]: It's not merged yet, added a TODO.

File: storage/mysql/storage_test.go (hunk 2 above)
Comment (codingllama) [inline storage/mysql/storage_test.go]: As discussed offline, I'll add a TODO for now.

File: storage/mysql/storage_test.go (hunk 4 above)
Comment (codingllama) [inline storage/mysql/storage_test.go]: As discussed offline, I'll add a TODO for now.

File: storage/mysql/storage_test.go (hunk 3 above)
Comment (codingllama) [inline storage/mysql/storage_test.go]: Added a TODO.

**Comment by gdbelvin** [comment]
These changes look great.

**Comment by codingllama** [comment]
Thanks, merging!

**Comment by daviddrysdale** [comment]
If I run the log integration test there seem to be an awful lot of calls to `defaultRegistry.GetLogStorage()` -- is this expected?

(If I add a line to that function with `log.Infof("Creating new storage for log: %d", treeID)` and turn on `--alsologtostderr` for the `trillian_log_server` invocation in `integration/log_integration_test.sh`, then running the test hits it ~850 times.)

**Comment by gdbelvin** [comment]
GetLogStorage currently takes the LogID as an input parameter, forcing
calls to GetLogStorage to occur within each RPC.
So at the moment, this is expected, but not desirable behavior.  Should be
fixed when https://github.com/google/trillian/issues/324 is addressed.

**Comment by daviddrysdale** [comment]
OK, thanks for checking.

**Pull request merged**